File: 19-08-2024_Kth_Smallest.py
class Solution:
    
    def show(self, arr, i, n):
        l = i * 2 + 1
        r = i * 2 + 2
        max_idx = i

        if r < n:
            if arr[l] > arr[r]:
                max_idx = l
            else:
                max_idx = r
        elif l < n:
            max_idx = l

        if arr[max_idx] > arr[i]:
            arr[i], arr[max_idx] = arr[max_idx], arr[i]
            self.show(arr, max_idx, n)
    
    def kthSmallest(self,arr, l, r, k):
        for i in range((r + 1) // 2 - 1, -1, -1):
            self.show(arr, i, r + 1)

        for i in range(r, k - 1, -1):
            arr[0], arr[i] = arr[i], arr[0]
            self.show(arr, 0, i)

        return arr[0]

File: 19-08-2024_Kth_Smallest.cpp
class Solution{
    public:
    // arr : given array
    // l : starting index of the array i.e 0
    // r : ending index of the array i.e size-1
    // k : find kth smallest element and return using this function
 //Using Max Heap Sort
void show(int arr[],int i,int n)
{
  int l=i*2+1; 
  int r=i*2+2;
  int t;
  int max=i;
  if(r<n)
  {
      if(arr[l]>arr[r])
      {
        max=l;  
      }
      else
      {
          max=r;
      }
  }
  else if(l<n)
  {
      max=l;
  }
  if(arr[max]>arr[i])
  {
      t=arr[i];
      arr[i]=arr[max];
      arr[max]=t;
      show(arr,max,n);
  }
}

int kthSmallest(int arr[], int l, int r, int k)
{
    for(int i=r+1/2-1;i>=0;i--)
    {
        show(arr,i,r+1);
    }
           int t;
    for(int i=r;i>=k;i--)
    {
       t=arr[0];
       arr[0]=arr[i];
       arr[i]=t;
       show(arr,0,i);
    }
return arr[0];
}
};
